pkg/dns/dnsprovider/dnsProvider.go

```diff
@@ -0,0 +1,62 @@
+package dnsprovider
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/apis/v1alpha1"
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/dns"
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/dns/aws"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var errUnsupportedProvider = fmt.Errorf("provider type given is not supported")
+
+type providerFactory struct {
+	client.Client
+}
+
+func NewProvider(c client.Client) *providerFactory {
+
+	return &providerFactory{
+		Client: c,
+	}
+}
+
+// depending on the provider type given in dnsprovider cr using the credentials supplied returns a dnsprovider.
+func (p *providerFactory) DNSProviderFactory(ctx context.Context, managedZone *v1alpha1.ManagedZone) (dns.Provider, error) {
+	providerSecret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      managedZone.Spec.SecretRef.Name,
+			Namespace: managedZone.Spec.SecretRef.Namespace,
+		}}
+
+	if err := p.Client.Get(ctx, client.ObjectKeyFromObject(providerSecret), providerSecret); err != nil {
+		return nil, err
+	}
+
+	if providerSecret.Type != "kuadrant.io/aws" {
+		return nil, fmt.Errorf("no dnsprovider types were found")
```

Comment:
I don't think you really need this if you add a default in the switch statement below that throws an error.

Reply:
sure i can remove it, theres a default there do you suggest something different added?

Reply:
No, it's throwing `errUnsupportedProvider` so should be good. It will never throw that error with this extra check here.

Reply:
yes i understand makes sense